### PR TITLE
add mavsdk_server restart if already running at connection

### PR DESCRIPTION
**Pull request by NicolasM0**
I have a lot of issues trying to connect to my autopilot. Sometimes, it fail because the autopilot boot is not complete or serial port is locked elsewhere. If my connection fail, I cannot retry because mavsdk server is blocked in the error.

On connection, I added a restart of the server if it is already running and was launch by this instance of mavsdk-python. It allows me retry to connect.
It also gives an access to stop a server internally launched.

I am not sure this is a good way to do this. Are you ok with that or do you have suggestions ?

### CODE REVIEW

**Review by JonasVautherin — COMMENTED** [review]
I am all for it if there is a need. Question now is: why does `mavsdk_server` hang when the serial connection fails? Can't we solve that from C++ by just returning an error and stopping gracefully?

File: mavsdk/system.py [hunk 1]
```diff
@@ -63,12 +63,26 @@ async def connect(self, system_address=None):
                 - TCP: tcp://[server_host][:server_port]
 
         """
-        if self._mavsdk_server_address is None:
+
+        if self._server_process is not None:
+            self._server_process.kill()
+            import time
+            time.sleep(1)
+            self._server_process = self._start_mavsdk_server(system_address,self._port)
```
[the remark this answers is not in the document]
Comment (NicolasM0) [reply mavsdk/system.py:71]: when server is launched internally, `_mavsdk_server_address` is set to `localhost` so I cannot test `_mavsdk_server_address` to know if it is launched by python or just running on localhost.
Instead, I keep the Popen result to be able to control the process. Checking if Popen result stored in `self._server_process` seems to be a good way to know if the server is run internally.

**Comment by NicolasM0** [comment]
> I am all for it if there is a need. Question now is: why does `mavsdk_server` hang when the serial connection fails? Can't we solve that from C++ by just returning an error and stopping gracefully?

Yes that could be a good solution.

File: mavsdk/system.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (JonasVautherin) [reply mavsdk/system.py:71]: Oh, right, I misread `self._server_process is not None` for `self._mavsdk_server_address is not None`. My bad :+1:

**Comment by JonasVautherin** [comment]
> Yes that could be a good solution.

Since you can reproduce that error, I would be super glad if you could have a look at the C++ side. And if we don't find an easy solution there, then we can totally go for this solution :+1:. I'm just thinking that this wrong behavior in C++ will also affect other users in other languages :innocent:.

**Comment by NicolasM0** [comment]
> > Yes that could be a good solution.
> 
> Since you can reproduce that error, I would be super glad if you could have a look at the C++ side. And if we don't find an easy solution there, then we can totally go for this solution +1. I'm just thinking that this wrong behavior in C++ will also affect other users in other languages innocent.

I had a look on that. We can focus on I of my issues: reboot command.
When, using serial,  mavsdk send a reboot command, serial connection is lost (that is expected). Then I would like to reconnect mavsdk to the autopilot but I cannot. mavsdk does not try to reopen the serial port.
What is the expected behavior:
- Should I start a new mavsdk_server ? In this case, I think something similar to what I propose in this PR should be added.
- Should mavsdk auto reconnect to the autopilot ? I am not sure how to do that as the autopilot does not use the same port after reboot on Linux (I suppose this is due to mavsdk_server still using the previous one).
- Should mavsdk_server exit when serial port is closed ?
- Something else ?

**Comment by JonasVautherin** [comment]
Those are interesting points, thanks! To be honest, I don't have much experience with serial connections, so let's try to find something acceptable together :slightly_smiling_face:.

> Should I start a new mavsdk_server ? In this case, I think something similar to what I propose in this PR should be added.

I think that you will have the same issue here: you don't know the serial port to be used for the new connection, right?

> Should mavsdk auto reconnect to the autopilot ?

I seems a bit tricky here: say mavsdk is connected over a SiK radio. That's a serial connection to the SiK radio, but it could be interrupted (i.e. we could lose the HEARTBEAT messages), while the serial connection is actually still valid. In this case, we don't want to disconnect, we just want to wait for new heartbeats.

But if we disconnect and reconnect the SiK radio from the computer (i.e. unplug and replug the USB), then that may create a new serial device, and mavsdk probably cannot reliably guess that.

Question: would there be a way to detect, from userspace, that the serial device disappeared? Even if that's just working on a desktop Linux, that would help for development. But if that's not possible, I'm afraid that maybe the solution is to manually restart `mavsdk_server`, or maybe have some kind of custom script (outside from `mavsdk_server` that would do just that on your setup).

What do you think?

**Comment by NicolasM0** [comment]
> Those are interesting points, thanks! To be honest, I don't have much experience with serial connections, so let's try to find something acceptable together slightly_smiling_face.

I don't have experience with serial connection neither...
 
> > Should I start a new mavsdk_server ? In this case, I think something similar to what I propose in this PR should be added.
> 
> I think that you will have the same issue here: you don't know the serial port to be used for the new connection, right?

That is not a problem: In my python program, if I can start a new mavsdk_server, I can search a device in serial connections, try all serials while I haven't found the one a want, ask the user to choose, etc.  

> > Should mavsdk auto reconnect to the autopilot ?
> 
> I seems a bit tricky here: say mavsdk is connected over a SiK radio. That's a serial connection to the SiK radio, but it could be interrupted (i.e. we could lose the HEARTBEAT messages), while the serial connection is actually still valid. In this case, we don't want to disconnect, we just want to wait for new heartbeats.
> 
> But if we disconnect and reconnect the SiK radio from the computer (i.e. unplug and replug the USB), then that may create a new serial device, and mavsdk probably cannot reliably guess that.
> 
> Question: would there be a way to detect, from userspace, that the serial device disappeared? Even if that's just working on a desktop Linux, that would help for development. But if that's not possible, I'm afraid that maybe the solution is to manually restart `mavsdk_server`, or maybe have some kind of custom script (outside from `mavsdk_server` that would do just that on your setup).

I think that is possible. I don't have sik radio right now to be verify we can distinguish radio disconnection and a serial disconnection but I think that ok and I will check when I can.

It's ok for me to use a custom script to manage mavsdk_server but, just to understand, as mavsdk-python is able create a  mavsdk_server in a  thread, what would be the drawback in allowing mavsdk-python to manage this instance of mavsdk ?

**Comment by JonasVautherin** [comment]
> what would be the drawback in allowing mavsdk-python to manage this instance of mavsdk 

No I think it makes sense, I was first hoping we could do something nicer at the C++ level, but it seems more complicated than I thought. Let's go with your changes!

File: mavsdk/system.py [hunk 2]
```diff
@@ -260,3 +274,5 @@ def cleanup():
             p.kill()
 
         atexit.register(cleanup)
+
+        return p
```
Comment (JonasVautherin) [inline mavsdk/system.py:278]: It seems like this is never used. Or is it?

Comment (NicolasM0) [reply mavsdk/system.py:283]: it is used here: https://github.com/mavlink/MAVSDK-Python/pull/262/files#diff-533e5a3c2350d6297f370a3a339084b7065e00661f66016b08fe4d0585130262R78

It allow to retrieve the created thread to manage it later

**Comment by NicolasM0** [comment]
tested with this script 
```
import asyncio
from mavsdk import System
import time;
import serial.tools.list_ports

system_address = "serial://" + serial.tools.list_ports.comports()[0].device;

drone = System()
asyncio.run(drone.connect(system_address))

while True:
    time.sleep(1)

    try:
        asyncio.run(drone.action.reboot())
    except:
        pass

    time.sleep(10)
    system_address = "serial://" + serial.tools.list_ports.comports()[0].device;

    asyncio.run(drone.connect(system_address))
```
My autopilot reboot and reconnect endlessly

**Comment by NicolasM0** [comment]
> > what would be the drawback in allowing mavsdk-python to manage this instance of mavsdk
> 
> No I think it makes sense, I was first hoping we could do something nicer at the C++ level, but it seems more complicated than I thought. Let's go with your changes!

I will try to improve the serial loss issue in C++ but not will not be able to reconnect so I think this will be helpful

**Review by JonasVautherin — APPROVED** [review]

**Review by julianoes — APPROVED** [review]
Seems good, thanks!

**Pull request merged**